### PR TITLE
Support updating config values when they change

### PR DESCRIPTION
**Pull request by jcw-**
# Summary
This PR contains an update to the refresh flows to support a more consistent approach to handling refreshes triggered from either the user or a config change.

# Details

- All four config values (not just `baseFolders`, but also `ignoredFolders`, `maxDepth`, and `cacheProjectsBetweenSessions`) stored in the abstract class now get updated on the appropriate instance if they are modified
- A common refresh flow is now invoked regardless if it's a user or config change triggering it.
- The `ProjectProvider` is always refreshed if it's a user-initiated refresh
- Otherwise, the `ProjectProvider` is only refreshed if an underlying config change in the abstract class was detected.

# Notes
- The config comparison was moved into the abstract class to allow the appropriate config to be compared without having to switch on instance type
- I put the three derived instances into an array to make it easier to invoke things on them in a consistent way, another place that could leverage this same refactor usage would be the `getProjects()` methods to be generic using this same approach

# References
#139

### CODE REVIEW

**Comment by alefragnani** [comment]
Hi @jcw- ,

First of all, thanks for your PR 👍 

I'm finishing and update in the extension right now, for multi-root support, and once I release this update, I will take a look at your PR, to incorporate it in the main branch.

Thanks again

File: src/abstractLocator.ts [hunk 1]
```diff
@@ -226,4 +213,59 @@ export abstract class AbstractLocator {
...
+        let currentValue = null;
+
+        currentValue = config.get<string[]>(this.getKind() + ".baseFolders");
+        if (!this.arraysAreEquals(this.baseFolders, currentValue)) {
+            this.baseFolders = currentValue;
+            refreshedSomething = true;
+        }
+
+        currentValue = config.get<string[]>(this.getKind() + ".ignoredFolders", []);
+        if (!this.arraysAreEquals(this.baseFolders, currentValue)) {
```
Comment (alefragnani) [inline src/abstractLocator.ts]: At this point, the `currentValue` should be checked against `this.ignoredFolders`, not `this.baseFolder`. I guest it was just a _copy/paste_ issue.

File: src/extension.ts [hunk 2]
```diff
@@ -79,7 +80,7 @@ export function activate(context: vscode.ExtensionContext) {
     });
 
     context.subscriptions.push(vscode.workspace.onDidChangeConfiguration(cfg => {
```
Comment (alefragnani) [inline src/extension.ts]: The `onDidChangeConfiguration` event now uses a new event interface (`ConfigurationChangeEvent`) which has a `affectsConfiguration` function to check for changes on specific configurations. Using that should avoid using compare with _oldValues_.  Maybe adding something like a `needsRefresh` function in `AbstractLocator`, which would worry only for its configs.

**Comment by alefragnani** [comment]
I ended uip making the changes myself and merged with that 😄 
Thanks again for your PR

**Comment by jcw-** [comment]
Awesome, glad you were able to put it to use! I use this extension pretty much every day. :)

**Pull request merged**